src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Source/AnalyzeDiagnostics.cs

```diff
@@ -0,0 +1,139 @@
+        {
+            if (analysis is null)
+            {
+                throw new ArgumentNullException(nameof(analysis));
+            }
+
+            var context = analysis.UpgradeContext;
+            var projects = context.Projects.ToList();
+
+            foreach (var project in projects)
+            {
+                await GetDiagnosticsAsync(project, token).ConfigureAwait(false);
+
+                yield return new()
+                {
+                    Results = ProcessDiagnostics(),
+                };
+            }
+        }
+
+        private HashSet<ResultObj> ProcessDiagnostics()
+        {
+            var results = new HashSet<ResultObj>();
+            foreach (var diag in Diagnostics)
+            {
+                results.Add(new()
+                {
+                    // Diagnostic line numbers are zero-based so offsetting by one
+                    LineNumber = diag.Location.GetLineSpan().Span.End.Line + 1,
+                    FileLocation = diag.Location.GetLineSpan().Path,
+                    ResultMessage = diag.Descriptor.Description.ToString(System.Globalization.CultureInfo.InvariantCulture),
+                });
+            }
+
+            return results;
+        }
+
+        public async Task GetDiagnosticsAsync(IProject project, CancellationToken token)
+        {
+            if (project is null)
+            {
+                _logger.LogWarning("No project available.");
+                return;
+            }
+
+            var roslynProject = project.GetRoslynProject();
+
+            if (roslynProject is null)
+            {
+                _logger.LogWarning("No project available.");
+                return;
+            }
+
+            _logger.LogInformation("Running analyzers on {ProjectName}", roslynProject.Name);
+
+            // Compile with analyzers enabled
+            var applicableAnalyzers = await GetApplicableAnalyzersAsync(_allAnalyzers, project).ToListAsync(token).ConfigureAwait(false);
+
+            if (!applicableAnalyzers.Any())
+            {
+                Diagnostics = Enumerable.Empty<Diagnostic>();
+            }
+            else
+            {
+                var compilation = await roslynProject.GetCompilationAsync(token).ConfigureAwait(false);
+                if (compilation is null)
+                {
+                    Diagnostics = Enumerable.Empty<Diagnostic>();
+                }
+                else
+                {
+                    var compilationWithAnalyzer = compilation
+                        .WithAnalyzers(ImmutableArray.CreateRange(applicableAnalyzers), new CompilationWithAnalyzersOptions(new AnalyzerOptions(_additionalTexts), ProcessAnalyzerException, true, true));
+
+                    // Find all diagnostics that registered analyzers produce
+                    Diagnostics = (await compilationWithAnalyzer.GetAnalyzerDiagnosticsAsync(token).ConfigureAwait(false)).Where(d => d.Location.IsInSource);
```

Comment:
Why keep state? Why not return this as the diagnostics used?

Reply:
I was following the pattern in the SourceUpdaterStep.cs

Reply:
Changed